Overhaul of SegmentPullers to add consistency and retries

drcrallen opened the pull request:
The Goal of this is to make the segment pullers a little more general use. Specifically, this uses URIs when possible. URIs have a standard defined format (unlike the free-form loadSpec used previously). This also unifies a lot of the way segments are pulled, so that any datasource should be able to support a zip or gz request in a more standard manner.

General workflow goes like this:

1. LoadSpec makes sure the correct Puller is called with the correct parameters. This offloads the "standard" interface from the puller to the LoadSpec. The only thing segment information was being used for in the puller was to get the loadSpec, so I simply moved where the guaranteed interface resides since we have an actual LoadSpec now.
2. The Puller sets up general information like how to make an InputStream, how to find a file name (for .gz files for example), and when to retry. 
3. CompressionUtils does most of the heavy lifting when it can

* ~~Requires https://github.com/druid-io/druid-api/pull/37~~ Merged
* ~~Requires https://github.com/metamx/java-util/pull/22~~ Merged


~~I also want to try this on some S3 segments just to make sure it didn't break anything unexpected. But would like feedback on the approach first.~~ Tested, seems to be fine.

~~This *MIGHT* be subject to http://bugs.java.com/bugdatabase/view_bug.do?bug_id=7036144  Added [unit tests](https://github.com/metamx/java-util/blob/CompressionFileUtils/src/test/java/com/metamx/common/CompressionUtilsTest.java) Adding more to double-check.~~ Fixed with workaround in java-util

drcrallen commented on server/src/main/java/io/druid/segment/loading/OmniSegmentLoader.java:
```diff
@@ -127,22 +136,27 @@ public File getSegmentFiles(DataSegment segment) throws SegmentLoadingException
       }
 
-
-      loc.addSegment(segment);
```
Was this double-adding segments?

drcrallen replied:
Looked deeper, I think it is a bit redundant but doesn't double-account for segments.

drcrallen commented:
@fjy :  Does this work better for your polymorphic desires?

drcrallen commented:
I might be missing a binding somewhere, I'll do more testing later this week, but that won't change the overall approach.

xvrl commented:
can we provide a little bit more context in the description as to what we are trying to achieve and why this is needed? This may not be obvious to an outsider.

drcrallen commented:
@xvrl : updated, please let me know if that is more clear.

cheddar commented:
The reason for the separation as it currently exists is to give implementations the ability to define

1. How they want to store data on their given underlying storage layer
2. The best way to get at the data that is stored in the underlying storage layer 

The reason for the jankiness in the current implementation is because it is mid-migration.  I agree that it needs to be improved, but I think that the implementation changes done here lead to less flexbility and a more difficult implementation.

Specifically, if we build an API around URIs, then we are essentially forcing everyone to come up with a mechanism of storing all of the data they might need/want to store inside of a URI, where working with just JSON objects and Java fields can be a lot simpler.

IIRC, the specific problem that needed to be resolved in order to really do this section of code "right" was the DI layer.  Now that we have Guice injection integrated with Jackson, we should be able to do a really clean API that gives control over to the person integrating with whatever storage they are integrating with.

Specifically, I'd propose that the LoadSpec API should be

```
public interface LoadSpec {
  public void get(File outDir)
}
```

That is, you pass it a directory and it stores the files in there.  For our current implementations, the implementation of this loadSpec would essentially get the metadata as well as the relevant "Puller" injected and use those.  For future implementations (or things that are amenable to URI/URL semantics), there can be an implementation of this interface which specializes to utilizing URIs instead.

This 

1. Meets the original requirements of giving implementors full control over how they are storing things
2. Works around issues with things like what if you want to retry a failure but don't have a Throwable?  With the API change made in the druid-api library, I now have to create a brand new Throwable, count on the outer logic to catch that and pass it off to my Predicate so that I can choose to retry.  That's leaking a lot in the abstraction.
3. Completely eliminates the `DataSegment` argument from the loading interface and generifies it to something that should be usable for other storage types as well.

What do you think?

drcrallen commented:
One reason I've been pushing back against the "just give LoadSpec a directory" approach is that at its core ends up being "Let's not use a standard resource identification schema, let's make our own non-standard schema" which is something I'd like to avoid.

There are two parts of key points of data encoded in the loadSpec aside from the individual data identifier: 1. The filesystem, 2: the data container (gz, zip, etc)

I understand the desire to have a simple interface, but with this approach I foresee 3 options:

1. Each possible way of loading storing data has its own LoadSpec for each underlying file system (zip, gz, directory, tar.gz , whatever_other_format X hdfs, s3, file, cassandra)
2. Each filesystem LoadSpec handles its own file loading in completely non-consistent ways (what we have now) with regards to retries and what formats are supported.
3. Each possible Filesystem supplies a handle to the data container, and the OmniSegmentLoader figures out how to handle that container properly (what I have proposed)

Obviously I'm partial to 3, but I'm ok with going route 1 above, even though I think its unclean compared to 3, but am very against 2.

cheddar commented:
So, I don't see an argument against 2 aside from you being very against it.  Why is it bad to allow implementations to control the "best" way to store the data and subsequently retrieve it?

Also, why is it bad to allow, for example, HDFS to automatically retry things in its client while s3 doesn't implement that for us so we have to implement it ourselves?  If you have something that does retries on its own already and you implement retries on top of it, you are now multiplying out the retries.

On the question of using a URI, your argument is essentially that a URI is a standard thing, so therefore we should take all of the things in the JSON fields and map them into a URI and then write code to map out of that URI back into JSON fields.  Fwiw, JSON is also a standard and we already have stuff setup to serialize into and out of it.  The fact that the switch to URIs is the equivalent of just adding a new marshalling layer to the code is very clear from looking at the Cassandra implementation (one big glaring thing is the fact that you are choosing not to implement the URI-based method...).

Note, I'm not against using URIs for things that already leverage URIs and I'm also not against having something that is intelligent about various storage formats at the end of those URIs.  I *am* against making the assumption that everything must fit that mold forever and always.  If you adopt my interface, the Cassandra stuff can implement that and you can have all of your logic and figuring out of things hidden inside of another concrete implementation of LoadSpec that leverages URIs.  That's perfectly fine.  If we adopt your changes, we do not have that flexibility.

drcrallen commented:
The reason I'm against the second option above is because there is a complete lack of consistency in how any data is handled. I *am* proposing using URI as a standard marshaling schema for resource identification because it is a pretty universally used resource identification schema. Json is a standard for data transfer, but to my knowledge there is no standard for resource identification with regards to Json.

Regarding why retries at the OmniSegmentLoader level: 1. It forces the developer to address retries as an inherent desire for the data. 2. Retry loops at the OmniSegmentLoader level are still controllable by the Puller via the retry predicate.

Regarding "Why is it bad to allow implementations to control the "best" way to store the data and subsequently retrieve it?" It goes back to consistency and *what level we want consistency to be enforced*. I was under the impression that there should be a *consistency in handling of data containers*, and the filesystem should be the pluggable portion. From most of your arguments I'm gathering that you feel the only consistency should be that a *pull to directory* can be called, and if any LoadSpec has to handle that in either wildly different or completely code-copied ways, then that's how its going to be. Please correct me if that's wrong.

cheddar commented:
@drcrallen Yes, I believe you are correct in capturing my thoughts on the matter.  A little bit of extra information is that I was thinking that exposing things in reusable code snippets and allowing the implementations to use them would be beneficial (hence the "CompressionUtils") in lowering the burden of repeated implementations.  

That said, I'm not against having an injection point that just implements an FS, allowing for the plugging that you are talking about.  I would prefer that both exist (and the docs could generally point people to the FS one and we reserve conversations on the mailing list for telling people about the other one).

One big reason for preferring both is that the way you have changed things right now, the change cannot be made backwards compatible.  I'm not sure if anyone is currently adding their own implementation as an extension (there was someone who implemented something on RackSpace that I'm not sure of the status of), but if someone is and they update their version of druid, all of a sudden their code won't work.

If we adjust the internal interfaces in the way I proposed, we can 

1. Have a backwards compatible implementation that uses the old interface
2. Introduce your new interface (with a different name) and have an implementation that uses that
3. Implement the Cassandra loader in terms of my proposed interface

This gives us the simplicity of just implementing an FS for the cases where that can be done easily, keeps an injection point for uses that don't lend themselves to URIs as well, and allows us to maintain backwards compatibility with old interface implementations, which can be deprecated and allow us a forward migration strategy.

I.e. hopefully it's a win for everyone and it should clean up various parts of the code where you are checking for URIs and such (the URI implementation can just be pure URI, it doesn't have to be dirtied with legacy support.

drcrallen commented:
@cheddar : I'm at mesos hackweek this week, so my time to modify this is limited, but I think unloading some of the brute work to a combination of FileUtils / CompressionUtils / RetryUtils is palatable to me. So the overall statement would be like this:

1. LoadSpec would be a (hopefully thin) helper to make sure the right combination of FileUtils / CompressionUtils / RetryUtils and the correct Puller are called for the given load type. 
2. LoadSpec simply guarantees that a destination directory be passed in as a File type
3. Any new file system for deep storage would need to make sure that a LoadSpec and the appropriate puller/pusher/killer was available
4. There are no guarantees as to what types of file/data containers each filesystem can support. Support for containers exists in CompressionUtils.

Does that sound like a reasonable solution statement?

cheddar commented:
Yes, that sounds reasonable to me.  Also, if you want to preserve working through URIs, I think a LoadSpec implementation based on loading a URI can make sense.

drcrallen commented:
@cheddar : Updated. Can you please take a look at the updated PR description and code and see if that fits in better with your workflow desires?

drcrallen commented on common/src/main/java/com/fasterxml/jackson/databind/introspect/GuiceInjectableValues.java:
```diff
@@ -36,6 +37,17 @@ public Object findInjectableValue(
   )
   {
-    return injector.getInstance((Key) valueId);
+    if(Key.class.isAssignableFrom(valueId.getClass())){
```
This was a weird error I was getting where sometimes beanInstance was being passed as String of the class rather than a Key of the class.

xvrl replied:
This may not be that odd, the documentation for `InjectableValues` states that `valueId` can be an *"Object that identifies value to inject; may be a simple name or more complex identifier object, whatever provider needs"*

It might make sense to add a comment though as to why this check is needed.

xvrl replied:
also, instanceof Key / String should be enough here, no?

drcrallen commented on extensions/cassandra-storage/src/main/java/io/druid/storage/cassandra/CassandraDruidModule.java:
```diff
@@ -35,18 +35,21 @@
+           .to(CassandraDataSegmentPuller.class)
+           .in(LazySingleton.class);
+
+    Binders.dataSegmentPullerBinder(binder)
```
This shouldn't be in here twice, will fix.

xvrl commented on common/src/main/java/com/fasterxml/jackson/databind/introspect/GuiceInjectableValues.java:
```diff
@@ -36,6 +37,17 @@ public Object findInjectableValue(
+    }
+    if(String.class.isAssignableFrom(valueId.getClass())){
+      try {
+        return injector.getInstance(getClass().getClassLoader().loadClass((String)valueId));
```
do we know which classes require a string valueId? It seems strange to have to resort to loading the class here, given that we don't necessarily know which classloader we should be using here.

drcrallen replied:
I think this was leftover from some of my advanced Jackson handling tests, which didn't end up working out. Will revert

cheddar commented on extensions/cassandra-storage/src/main/java/io/druid/storage/cassandra/CassandraDataSegmentPuller.java:
```diff
@@ -26,70 +26,76 @@
+    long startTime = System.currentTimeMillis();
+    ObjectMetadata meta = null;
+    final File tmpFile = new File(outDir, "index.zip");
+    tmpFile.deleteOnExit();
```
`tmpFile.deleteOnExit()` registers the file with a shutdown listener that I'm pretty sure will exist until the process dies.  I.e. for a really long-running process, this is a memory leak.  It might be a slow trickle, but still likely a leak.  Can we make sure we delete the thing using a finally block or something instead?

drcrallen replied:
See line 96. This is just a safeguard to keep from polluting the filesystem in case something goes really wacky.

drcrallen replied:
I removed this anyways to make sure there's not memory pollution. The delete in `finally` is still there.

cheddar commented on extensions/cassandra-storage/src/main/java/io/druid/storage/cassandra/CassandraLoadSpec.java:
```diff
@@ -0,0 +1,65 @@
+  }
+
+  @Override
+  public Long putSegment(File outDir) throws SegmentLoadingException
```
the name `putSegment` makes me imagine pushing to storage.  What do you think about `loadSegment`?

drcrallen replied:
I'm fine with that. Will change on next pass

drcrallen replied:
Fixed

cheddar commented on extensions/cassandra-storage/src/main/java/io/druid/storage/cassandra/CassandraLoadSpec.java:
```diff
@@ -0,0 +1,65 @@
+  private final CassandraDataSegmentPuller puller;
+  @JsonCreator
+  public CassandraLoadSpec(
+      @JacksonInject Map<String, DataSegmentPuller> pullers,
```
Why not just inject `CassandraDataSegmentPuller`?

drcrallen replied:
Early attempts were to separate where the puller and dependencies must be available. I'll need to double-check the logic here, but the original goal was to make it so that a node (like the coordinator) could read an hdfs load spec without needing to have the hdfs extension. I probably need to move the puller assignment into the interface function, that way you don't actually have to have the pullers available unless you're trying to use the functionality of the interface. 

This same comment applies to the other LoadSpecs

drcrallen replied:
The LoadSpec is stored as a Map until the OmniSegmentLoader is called to load the segment. At that point, the LoadSpec's respective modules have to be loaded anyways, so the mapper injection can go away in favor of the explicit puller injection.

cheddar commented on extensions/hdfs-storage/src/main/java/io/druid/storage/hdfs/HdfsDataSegmentPuller.java:
```diff
@@ -42,25 +156,131 @@ public HdfsDataSegmentPuller(final Configuration config)
+              {
+                return getInputStream(path);
+              }
+            }, outDir, shouldRetryPredicate(), false
```
It appears like retrying is being lumped into the CompressionUtils?

How would you feel about having a "RetryUtil" or something that takes a `Callable<RetryReturn<T>>`, where `RetryReturn` is essentially a response saying "retry me" or holds a reference to some return object?  The callable is called and if it returns something that says "retry me" the RetryUtil determines if it has retried enough to give up or not and, if it hasn't, it runs it again.  If it has, it just gives up and returns a null?

drcrallen replied:
That's essentially what RetryUtils does. Under the hood the compression utils eventually just uses the `FileUtils.retryCopy` https://github.com/metamx/java-util/pull/22/files#diff-9ff99574eeaa8825730a16d0d4edee1dR104

drcrallen replied:
the MAJOR difference being that an exception is thrown instead of just returning null (this was the prior behavior of RetryUtils and I did not modify that behavior)

cheddar commented on extensions/hdfs-storage/src/main/java/io/druid/storage/hdfs/HdfsLoadSpec.java:
```diff
@@ -0,0 +1,65 @@
+  final HdfsDataSegmentPuller puller;
+  @JsonCreator
+  public HdfsLoadSpec(
+      @JacksonInject Map<String, DataSegmentPuller> pullers,
```
Same question as before, why not just inject the HdfsDataSegmentPuller directly?

Also, it might be useful to put the `SegmentPuller` injection into a setter so that it's easier to generate load specs just to serialize them out.

cheddar replied:
Also, probably want to add an `@Nullable`

drcrallen replied:
I changed this to inject the puller directly instead of the puller map. I went against making it `@Nullable` because the LoadSpec remains a `Map<String, Object>` until it is immediately ready for consumption. Anywhere that actually consumes a LoadSpec also has to have the module loaded that the LoadSpec is a part of. This means that anything that will actually consume a HdfsLoadSpec must have the HDFS storage module added, and the same for the S3 and Cassandra modules, so I'm proposing  the rule:

Don't materialize a Map<String, Object> into a LoadSpec unless you're ready to consume it.

Which alleviates dependency issues because anyone who is ready to consume it sure better have the correct module loaded.

cheddar commented on extensions/s3-extensions/src/main/java/io/druid/storage/s3/S3LoadSpec.java:
```diff
@@ -0,0 +1,94 @@
+
+  @JsonCreator
+  public S3LoadSpec(
+      @JacksonInject Map<String, DataSegmentPuller> pullers,
```
same general comments here as on Hdfs and Cassandra

cheddar commented on server/src/main/java/io/druid/segment/loading/OmniSegmentLoader.java:
```diff
@@ -52,12 +54,14 @@
   public OmniSegmentLoader(
```
Hrm, I was actually imagining this going away.  The job of the OmniSegmentLoader was essentially to do the polymorphism that we should be getting out of `LoadSpec`s now.  It does some extra stuff, but that should be able to be lifted to the place that uses the LoadSpec.

That is, when the metadata gets deserialized, that can have a LoadSpec object in it which is just called to load the given segment.

drcrallen replied:
Right now one of the functions of omni-segment-loader is to have an explicit point where the dependencies must be materialized. There is a place where the LoadSpec's method is called, and that is the point where the dependency extensions must be loaded and available.

drcrallen replied:
I'll re-look at it, but I think OmniSegmentLoader still provides some functionality. I'll get a writeup in a bit on if it is still needed.

drcrallen replied:
Currently the OmniSegmentLoader handles a lot of functionality around cache management, so things like:

1. Manage choosing index cache locations
1. Manage cleaning up index cache locations
1. Manage download coordination of caches (downloadMarker)
1. Primary implementation of `io.druid.segment.loading.SegmentLoader`

So effectively, the OmniSegmentLoader becomes the "SegmentLocalCacheManager"

cheddar replied:
Having it become the `SegmentLocalCacheManager` sounds reasonable to me.  It definitely clarifies what it is actually doing.

drcrallen replied:
renamed to `SegmentLoaderLocalCacheManager`

cheddar commented on extensions/hdfs-storage/src/main/java/io/druid/storage/hdfs/HdfsLoadSpec.java:
```diff
@@ -0,0 +1,65 @@
+  }
+
+  @Override
+  public Long putSegment(File outDir) throws SegmentLoadingException
```
Given that this is a public API, let's give this a proper object return with a getter for `size`.  This will make it easier to add return parameters to the method in the future in a forwards compatible manner.

drcrallen replied:
Do we have any objects already that would fit this use case well?

drcrallen replied:
DataSegment exists, that has similar functionality to what you describe, but this is a servant to DataSegment.

Originally this effectively returned Void. This was changed to Long to allow some level of sanity checking in the future. I'm trying to think of things that would be helpful for DataSegment (or its handler) to know that isn't already in DataSegment. The only thing I could think of would be some sort of handler to the files that were loaded (so like a thin wrapper for `Iterable<File>` or something) then whatever is handling the DataSegment could use that info in whatever way it needs.

The LoadSpec pretty much only knows "where do I get files" and "how do I tell the puller the correct place to get them and the correct place to put them", so its knowledge realm is pretty thin.

drcrallen commented:
@cheddar  : Updated the file loading *partially*. Now java-util at least propagates an object which contains some extra information.

Going through, though I couldn't really get a handle on what information would be helpful to the pullers or omni-segment-loader other than the size of data written to disk. If you can think of any particular abstraction that would be helpful, please let me know, otherwise I can simply make a pojo containing the size and use that if you want at least some level of future proofing.

drcrallen commented:
@cheddar : Added a basic datatype to the return of the method in LoadSpec

Also added basic sanity check to the SegmentLoader which checks the expected vs the written size of the segments and logs a warning.

fjy commented on extensions/hdfs-storage/pom.xml:
```diff
@@ -63,13 +63,39 @@
+        </dependency>
+        <dependency>
+          <groupId>org.apache.hadoop</groupId>
+          <artifactId>hadoop-hdfs</artifactId>
```
why are these required?

drcrallen replied:
`<scope>test</scope>`
for unit tests.

fjy commented on extensions/cassandra-storage/src/main/java/io/druid/storage/cassandra/CassandraDataSegmentPuller.java:
```diff
@@ -17,79 +17,106 @@
-	private static final int BATCH_SIZE = 10;
+  public static final String SCHEME = "cassandra";
+  private static final Logger log = new Logger(CassandraDataSegmentPuller.class);
+  private static final int CONCURRENCY = 10;
```
at some point if these were configurable it'd be nice

drcrallen replied:
I find it more likely that the Cassandra plugin will have a complete overhaul.

fjy commented on extensions/hdfs-storage/src/main/java/io/druid/storage/hdfs/HdfsDataSegmentPuller.java:
```diff
@@ -17,23 +17,145 @@
+    };
+  }
+
+  public static final String scheme = HdfsStorageDruidModule.SCHEME;
```
why is this needed?

drcrallen replied:
Fixed

cheddar commented:
Aside from a few "?"s (and me commenting on the commit instead of the PR... sorry), this looks generally in line with my comments.

drcrallen commented:
@cheddar : I had some "magic" logic like that previously related to some other stuff, and there was concern raised by @gianm  about it.

In this particular case the most concerning part is if it is gzip, would you want it to rename the file to the base name and output to the directory, or are you wanting to pass in a file you want the ungzip'd data to go to?

There is not obvious auto-magic solution in such a case without having flags which only pertain to the gzip case.

drcrallen commented:
Regarding eliminating `.gz` code: it is used in some of the QTL stuff.

This PR is a baby step towards giving a more generic and standard interface between data. If I could pull in hadoop's abstract File handling without encountering Hadoop dependency hell I would.

fjy commented on server/src/main/java/io/druid/segment/loading/LocalDataSegmentPuller.java:
```diff
@@ -17,58 +17,196 @@
+  @Override
+  public Predicate<Throwable> shouldRetryPredicate()
+  {
+    // TODO: smarter retries
```
can we leave a comment here about why smarter retries are necessary and remove the TODO?

drcrallen replied:
Fixed

fjy commented on server/src/main/java/io/druid/segment/loading/SegmentLoaderLocalCacheManager.java:
```diff
@@ -127,22 +127,27 @@ public File getSegmentFiles(DataSegment segment) throws SegmentLoadingException
         try {
-          downloadStartMarker.createNewFile();
+          if (!downloadStartMarker.createNewFile()) {
+            // TODO: should this fail?
```
i think we should trhow an exception here

drcrallen replied:
Fixed

cheddar commented:
On the decompress thing.  I think it can make sense to have the semantics be "give me the directory to decompress files into" and in the gzip case, it keeps the original file name without the `.gz` (similar to what the unix `gunzip` does).  For zip, it just unzips into the directory.

cheddar commented on server/src/test/java/io/druid/segment/loading/LoadSpecTest.java:
```diff
@@ -0,0 +1,143 @@
+ *
+ */
+@RunWith(Parameterized.class)
+public class LoadSpecTest
```
I'm not sure what this test is trying to verify exactly.  It looks like it's trying to verify that Jackson can instantiate an instance.  If that is truly the case, then I think there's a lot of extra stuff in the setup that doesn't really help.

At a minimum, I'd recommend just switching the `setup` to not bother with a Guice injector and just add the `GuiceAnnotationIntrospector` as well as an InjectableValues that checks to make sure that it gets the Key and then just returns null.

Also, when you register stuff with Jackson, you can just use a SimpleModule and call `registerSubType()`.

drcrallen replied:
I'll make some changes but the test is to verify that the minimal components needed to inject are well defined. It's more of a "Do I really understand all the injection dependencies" rather than "can Jackson do what Jackson does"

drcrallen replied:
Fixed

cheddar commented on server/src/test/java/io/druid/segment/loading/LoadSpecTest.java:
```diff
@@ -0,0 +1,143 @@
+            if (valueId instanceof Key) {
+              return injector.getInstance((Key) valueId);
+            }
+            if (valueId instanceof String) {
```
The reason you are getting a String here is because you don't have the GuiceAnnotationIntrospector registered with the ObjectMapper.

drcrallen replied:
Fixed

drcrallen commented:
@cheddar : Ping! Last I heard you were pretty good with this one overall but had some small outstanding comments on the PR. I think those have been addressed.

cheddar commented:
Seems fine to me
